src/main/java/org/dspace/pack/bagit/BagItAipReader.java

```diff
@@ -0,0 +1,293 @@
+            if (iterator.hasNext()) {
+                logo = Optional.of(iterator.next());
+            }
+        }
+
+        return logo;
+    }
+
+    /**
+     * Read the metadata.xml file located in a bags data directory
+     *
+     * @return a list of {@link XmlElement}s which were read from the file
+     * @throws IOException if there was an error reading the file or parsing the xml
+     */
+    public List<XmlElement> readMetadata() throws IOException {
+        final Path metadata = bag.resolve(metadataLocation);
+        try (InputStream metadataStream = Files.newInputStream(metadata)) {
+            return readXml(metadataStream);
+        }
+    }
+
+    /**
+     * Find any {@link org.dspace.content.Bitstream}s packaged in an aip and return path, bundle name, and metadata
+     *
+     * @return a list of {@link PackagedBitstream}s found
+     * @throws IOException if there are any errors searching for bitstreams
+     */
+    public List<PackagedBitstream> findBitstreams() throws IOException {
+        final Path data = bag.resolve(dataDirectory);
+
+        // build our regex
+        // matches bitstream_uuid OR bitstream_uuid.extension with a group for bitstream_uuid
+        final String bitstreamStart = "bitstream_";
+        final String relaxedUuid = "[\\w]{8}-[\\w]{4}-[\\w]{4}-[\\w]{4}-[\\w]{12}";
+        final String extension = "(\\..*)?";
+        final Pattern uuid = Pattern.compile("(?<uuid>" + bitstreamStart + relaxedUuid + ")" + extension);
+
+        // filter to find only directories (for bundle names)
+        final DirectoryStream.Filter<Path> directoryFilter = new DirectoryStream.Filter<Path>() {
+            @Override
+            public boolean accept(Path path) {
+                return Files.isDirectory(path);
+            }
+        };
+
+        // filter to find only the bitstream file
+        final DirectoryStream.Filter<Path> bitstreamFilter = new DirectoryStream.Filter<Path>() {
+            @Override
+            public boolean accept(Path path) {
+                return path.toFile().isFile() && uuid.matcher(path.getFileName().toString()).matches();
+            }
+        };
+
+        final List<PackagedBitstream> packagedBitstreams = new ArrayList<>();
+
+        // iterate all bundles
+        try (DirectoryStream<Path> directories = Files.newDirectoryStream(data, directoryFilter)) {
+            for (Path bundle : directories) {
+                final String bundleName = bundle.getFileName().toString();
+
+                // iterate all bitstreams
+                try(DirectoryStream<Path> bitstreams = Files.newDirectoryStream(bundle, bitstreamFilter)) {
+                    for (Path bitstream : bitstreams) {
+                        final String bitstreamName = bitstream.getFileName().toString();
+
+                        // load the bitstream metadata
+                        final Matcher matcher = uuid.matcher(bitstreamName);
+                        // todo: error on no match?
```

Comment:
Since the `DirectoryStream` filters against the same regex a non-match should never occur. Could still opt to throw an `AssertionError`, but likely is not necessary anymore.

Reply:
Is `findBitstreams()` called for all AIPs or just Items? All Items should have a bitstream, so not finding one on a restore would be an error (and an exception would make sense). I believe Communities and Collections can have a logo bitstream, but often do not, so if this is being used to process Communities and Collections as well, you don't want to throw an error if there are no bitstreams.

Reply:
`findBitstreams()` is called only for Items. I can see how there could be confusion between a logo and bundled bitstream, which could lead to confusion about usage of the different methods (i.e. someone wants to have the logo as a `PackagedBitstream` when that isn't possible), which I hadn't considered before.

I actually might try to remove all bitstreams from an Item and see what happens, that isn't really something I had thought of as an error case.